File: Go/GORM/utils/EntityService.go
package utils

import (
	"gorm/database"
	"gorm/models"
	"time"
)

type EntityService struct {
	dbFactory database.DbFactory
}

type EntityError string

func (e EntityError) Error() string {
	return string(e)
}

const EntryIsNullError EntityError = "entry cannot be null"

func NewEntityService(dbFactory database.DbFactory) EntityService {
	return EntityService{dbFactory: dbFactory}
}

func (s EntityService) get(entry interface{}, filter interface{}) error {
	if entry == nil {
		return EntryIsNullError
	}

	db := s.dbFactory.GetDB()
	return db.First(entry, filter).Error
}

func (s EntityService) GetDepartment(id uint) (dep *models.Department, err error) {
	dep = &models.Department{}
	filter := &models.Department{ID: id}
	err = s.get(dep, filter)
	return
}

func (s EntityService) GetPerson(id string) (per *models.Person, err error) {
	per = &models.Person{}
	filter := &models.Person{Id: id}
	err = s.get(per, filter)
	return
}

func (s EntityService) GetEmployee(id string) (emp *models.Employee, err error) {
	emp = &models.Employee{}
	filter := &models.Employee{Id: id}
	db := s.dbFactory.GetDB()
	err = db.Joins("Department").Joins("Person").First(emp, filter).Error
	return
}

func (s EntityService) addNew(entry interface{}) error {
	if entry == nil {
		return EntryIsNullError
	}

	db := s.dbFactory.GetDB()
	return db.Create(entry).Error
}

func (s EntityService) AddNewDepartment(name string) (dep *models.Department, err error) {
	dep = &models.Department{Name: name}
	err = s.addNew(dep)
	return
}

func (s EntityService) AddNewPerson(name, address string, birthDate time.Time) (per *models.Person, err error) {
	per = &models.Person{Name: name, Address: address, BirthDate: birthDate}
	err = s.addNew(per)
	return
}

func (s EntityService) AddNewEmployee(personId string, departmentId uint, position models.Position) (emp *models.Employee, err error) {
	per, err := s.GetPerson(personId)
	if err != nil {
		return
	}

	dep, err := s.GetDepartment(departmentId)
	if err != nil {
		return
	}

	tEmp := &models.Employee{PersonId: personId, DepartmentId: departmentId, Position: position}
	err = s.addNew(tEmp)
	if err != nil {
		return
	}

	emp = &models.Employee{
		Id:           tEmp.Id,
		PersonId:     personId,
		Person:       *per,
		DepartmentId: departmentId,
		Department:   *dep,
		Position:     position,
	}

	return
}

func (s EntityService) AddNewEmployeeAndPerson(departmentId uint, name, address string, birthDate time.Time, position models.Position) (emp *models.Employee, err error) {
	dep, err := s.GetDepartment(departmentId)
	if err != nil {
		return
	}

	per, err := s.AddNewPerson(name, address, birthDate)
	if err != nil {
		return
	}

	emp, err = s.AddNewEmployee(per.Id, dep.ID, position)
	return
}

func (s EntityService) Update(entry interface{}) error {
	if entry == nil {
		return EntryIsNullError
	}

	db := s.dbFactory.GetDB()
	return db.Save(entry).Error
}

func (s EntityService) GetAll(entries interface{}) error {
	if entries == nil {
		return EntryIsNullError
	}

	db := s.dbFactory.GetDB()
	return db.Find(entries).Error
}

func (s EntityService) Remove(entry interface{}) error {
	if entry == nil {
		return EntryIsNullError
	}

	db := s.dbFactory.GetDB()
	return db.Delete(entry).Error
}

func (s EntityService) RemoveRange(entries interface{}, ids []interface{}) error {
	if entries == nil {
		return EntryIsNullError
	}

	db := s.dbFactory.GetDB()
	return db.Delete(entries, ids).Error
}

File: PHP/Doctrine/src/Services/EntityService.php
<?php

namespace Services;

use DateTime;
use Doctrine\ORM\EntityManager;
use Models\Department;
use Models\Person;
use Models\Employee;
use Models\Position;

class EntityService
{
    public function __construct(private EntityManager $entityManager)
    {
        $this->entityManager = $entityManager;
    }

    private function get_entity(int | string $id, string $entity_name): mixed
    {
        $repo = $this->entityManager->getRepository($entity_name);
        $entity = $repo->find($id);
        return $entity;
    }

    public function get_department(int $id): Department
    {
        $entity_name = Department::class;

        /** @var Department */
        $dep = $this->get_entity($id, $entity_name);
        return $dep;
    }

    public function get_person(string $id): Person
    {
        $entity_name = Person::class;

        /** @var Person */
        $per = $this->get_entity($id, $entity_name);
        return $per;
    }

    public function get_employee(string $id): Employee
    {
        $entity_name = Employee::class;

        /** @var Employee */
        $emp = $this->get_entity($id, $entity_name);
        return $emp;
    }

    private function add_new(mixed $entity)
    {
        $this->entityManager->persist($entity);
        $this->entityManager->flush($entity);
    }

    public function add_new_department(string $name): Department
    {
        $dep = new Department($name);
        $this->add_new($dep);
        return $dep;
    }

    public function add_new_person(string $name, string $address, DateTime $birth_date): Person
    {
        $per = new Person($name, $address, $birth_date);
        $this->add_new($per);
        return $per;
    }

    public function add_new_employee(Person $per, Department $dep, Position $pos): Employee
    {
        $emp = new Employee($per, $dep, $pos);
        $this->add_new($emp);
        return $emp;
    }

    public function add_new_employee_and_person(Department $dep, string $name, string $address, DateTime $birth_date, Position $pos): Employee
    {
        $per = $this->add_new_person($name, $address, $birth_date);
        return $this->add_new_employee($per, $dep, $pos);
    }

    public function update_entity(mixed $entity)
    {
        $this->entityManager->flush($entity);
    }

    /**
     * @param string $entity_name
     * @return array<int, object>
     */
    private function get_all(string $entity_name): array
    {
        $rep = $this->entityManager->getRepository($entity_name);
        $entities = $rep->findAll();
        return $entities;
    }

    /**
     * @return array<int, Department>
     */
    public function get_all_departments(): array
    {
        return $this->get_all(Department::class);
    }

    /**
     * @return array<int, Person>
     */
    public function get_all_persons(): array
    {
        return $this->get_all(Person::class);
    }

    /**
     * @return array<int, Employee>
     */
    public function get_all_employee(): array
    {
        return $this->get_all(Employee::class);
    }

    public function remove(mixed $entity)
    {
        $this->entityManager->remove($entity);
        $this->entityManager->flush($entity);
    }

    public function remove_range(array $entities)
    {
        if ($entities && count($entities) == 0)
            return;

        foreach ($entities as $entity)
            $this->entityManager->remove($entity);

        $this->entityManager->flush($entities);
    }
}
